### PR TITLE
Add minio nas support with documentation and example

### PR DESCRIPTION
**Pull request by youngm**
Add support for minio nas such as NFS.

### CODE REVIEW

**Comment by harshavardhana** [comment]
Along these similar lines perhaps we should also add edge caching support in bosh releases as well. This might be super helpful to a lot of people.

**Comment by krishnasrinivas** [comment]
@youngm typically how would CF users mount a nfs volume? do they use nfs_mounter job? https://bosh.cloudfoundry.org/jobs/nfs_mounter?source=github.com/cloudfoundry/cf-release&version=198

**Comment by krishnasrinivas** [comment]
@youngm this PR looks good. Is there a way to ensure that nfs job mounts before minio-nas is started?

File: jobs/minio-nas/spec [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+properties:
+  credential.accesskey:
+    description: "The access key the server should accept"
+    default: ~
+  credential.secretkey:
+    description: "The secret key the server should accept"
+    default: ~
+  nas_data_dir:
+    description: "The directory where the nas share is mounted."
```
Comment (krishnasrinivas) [inline jobs/minio-nas/spec]: default can be `/var/vcap/nfs`

**Comment by youngm** [comment]
@krishnasrinivas That's a good question.  I mostly just hoped that minio in gateway mode would happily work if the volume was switched out from under it.  But if that is dangerous for minio I can do some more thinking about that and try to figure something out.

File: jobs/minio-nas/spec (hunk 1 above)
Comment (youngm) [reply jobs/minio-nas/spec]: I left it with no default because I didn't want to assume nfs or anything.  But I can add it if you want.

**Comment by youngm** [comment]
I supposed worst case if mounting failed we wouldn't want minio to assume it was working when in fact it was not.  I'll try and come up with some code to help detect if the data_dir is an external volume.

File: jobs/minio-nas/spec (hunk 1 above)
Comment (krishnasrinivas) [reply jobs/minio-nas/spec]: the nfs-mounter job from cf-release also used /var/vcap/nfs, but maybe it is a good idea to not assume anything and leave it blank.

Does bosh give error if a user does not provide this property in the deployment manifest? (if we do not have a default value in the spec)

**Comment by krishnasrinivas** [comment]
One option - if `stat --format="%d%D" /` != `stat --format="%d%D" /var/vcap/nfs` then we can assume that nfs is mounted

But ideally we want to establish dependencies. Can we use `consumes:` section on job spec to define dependencies?

**Comment by krishnasrinivas** [comment]
> Along these similar lines perhaps we should also add edge caching support in bosh releases as well. This might be super helpful to a lot of people.

Sure, we can add the cache-property to gateway's jobs @harshavardhana

File: jobs/minio-nas/spec (hunk 1 above)
Comment (youngm) [reply jobs/minio-nas/spec]: Yes.  Bosh will fail with a nice error message if it attempts to use a property in a template that doesn't exist.

**Comment by youngm** [comment]
> Can we use consumes: section on job spec to define dependencies?

No consumes and produces is only for sending and retrieving properties/config between jobs not determining order.  It is quite typical for a bosh job to fail to start and retry while waiting for a dependency to become available.  Worst case I think that would be fine.  I might be possible to add a monit health check that causes the job to fail if the share is not mounted.  I'll give it a few minutes and see how much I can improve the situation.

**Comment by youngm** [comment]
@krishnasrinivas PR updated with NAS mount check before minio starts.  I looked into a monit health check to detect if the NAS was up and it didn't seem super straightforward and the value of such a check seemed dubious.

Let me know if there is anything else you'd want in the PR.

File: jobs/minio-nas/templates/health_check.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+http_response=$(curl -H "User-Agent: Mozilla" -s -k -o /dev/null -I -w "%{http_code}" http://localhost:<%= p("port") %>/minio/index.html)
+
+# if the server is up it would return:
+# 200 - if minio-browser is enabled
+# 404 - if minio-browser is disabled
+[ "$http_response" = "200" ] || [ "$http_response" = "404" ]
```
Comment (harshavardhana) [inline jobs/minio-nas/templates/health_check.erb]: We have now readiness and check probes `/minio/health/ready` which can be used. https://github.com/minio/minio/tree/master/docs/healthcheck

Comment (krishnasrinivas) [reply jobs/minio-nas/templates/health_check.erb]: I have addressed the health check in this PR https://github.com/minio/minio-boshrelease/pull/60 which can be merged after the current PR is merged

**Comment by harshavardhana** [comment]
This is merged... thanks @youngm

**Pull request closed without merging**